pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -490,6 +493,46 @@ monitorLoop:
 	}
 }
 
+func (l *LibvirtDomainManager) CancelVMIMigration(vmi *v1.VirtualMachineInstance) error {
+	if vmi.Status.MigrationState == nil || vmi.Status.MigrationState.Completed ||
+		vmi.Status.MigrationState.Failed {
+
+		l.setMigrationResult(vmi, false, "", v1.MigrationAbortFailed)
+		return fmt.Errorf("failed to cancel migration - vmi is not migrating")
+	}
+	l.asyncMigrationAbort(vmi)
+	return nil
+}
+
+func (l *LibvirtDomainManager) asyncMigrationAbort(vmi *v1.VirtualMachineInstance) {
+	go func(l *LibvirtDomainManager, vmi *v1.VirtualMachineInstance) {
+
+		domName := api.VMINamespaceKeyFunc(vmi)
+		dom, err := l.virConn.LookupDomainByName(domName)
+		if err != nil {
+			log.Log.Object(vmi).Reason(err).Warning("failed to cancel migration, domain not found ")
+			l.setMigrationResult(vmi, false, "", v1.MigrationAbortFailed)
+			return
+		}
+		stats, err := dom.GetJobInfo()
+		if err != nil {
+			log.Log.Object(vmi).Reason(err).Error("failed to get domain job info")
+			l.setMigrationResult(vmi, false, "", v1.MigrationAbortFailed)
+			return
+		}
+		if stats.Type == libvirt.DOMAIN_JOB_UNBOUNDED {
+			err := dom.AbortJob()
+			if err != nil {
+				log.Log.Object(vmi).Reason(err).Error("failed to cancel migration")
+				l.setMigrationResult(vmi, false, "", v1.MigrationAbortFailed)
+				return
+			}
+			l.setMigrationResult(vmi, true, "Live migration aborted ", v1.MigrationAbortSucceeded)
```

Comment:
this function is called from multiple go routines. I believe they'll all clobber each others results.

For example, both asyncMigrate and this function attempt to record the migration result on the metadata. The migration failure reason might be different depending on who wrote it last.

Reply:
Yes, but I think the chance for the migration failing exactly at the same time as it's being canceled is very small.
Even if it did happen, the result will still be the same.

Reply:
Let's protect it with a channel or a semaphore and just call it once.

Reply:
I don't think that concurrency is what @davidvossel is concerned about, rather, one message can mask another. Posting only one doesn't solve anything, I think.

My argument is that it doesn't really matter at this point. The scenario is that the migration has failed exactly at the time of the cancellation.
If the migration has been canceled and we've posted the cancelation and the migration cashed and we posted the crash message after (I think this is not realistic) - but this doesn't matter, as the migration is not happening anymore and this is the original intended.
The other scenario is the migration crashed and the cancellation happened after - this cannot happen as the migration abort cannot succeed.

Reply:
Yeah, a mutex wouldn't help necessarily. Like we discussed out of band, having the monitor loop be the single go-routine that reports results should resolve my concern here.

Reply:
Done